CRM/Civioffice/DocumentRendererType/LocalUnoconv/PhpWordTemplateProcessor.php

```diff
@@ -13,124 +13,215 @@
+        // because it may contain HTML entities.
+        $this->setValue($macro_variable, $elements[0]->getText());
+      }
+      else {
+        // ... or as HTML: Render all elements and insert in the text
+        // run or paragraph containing the macro.
+        $this->setElementsValue($macro_variable, $elements);
+      }
+    }
+    catch (Exception $exception) {
+      throw new CRM_Core_Exception(
+        E::ts('Error loading/writing PhpWord document: %1', [1 => $exception->getMessage()]),
+        $exception->getCode(),
+        [],
+        $exception
+      );
+    }
+    finally {
+      PhpWord\Settings::setOutputEscapingEnabled($outputEscapingEnabled);
+    }
+  }
+
+  /**
+   * Replaces a search string (macro) with a set of rendered elements, splitting
+   * surrounding texts, text runs or paragraphs before and after the macro,
+   * depending on the types of elements to insert.
+   *
+   * @param string $search
+   * @param \PhpOffice\PhpWord\Element\AbstractElement[] $elements
+   *
+   * @return void
+   * @throws \CRM_Core_Exception
+   */
+  public function setElementsValue(string $search, array $elements) {
+    $elements_data = '';
+    $hasParagraphs = FALSE;
+    foreach ($elements as $element) {
+      $elementName = substr(
+        get_class($element),
+        (int) strrpos(get_class($element), '\\') + 1
+      );
+      $objectClass = 'PhpOffice\\PhpWord\\Writer\\Word2007\\Element\\' . $elementName;
+
+      // For inline elements, do not create a new paragraph.
+      $withParagraph = 'PhpOffice\\PhpWord\\Writer\\Word2007\\Element\\Text' !== $objectClass;
+      $hasParagraphs = $hasParagraphs || $withParagraph;
+
+      $xmlWriter = new PhpWord\Shared\XMLWriter();
+      /** @var \PhpOffice\PhpWord\Writer\Word2007\Element\AbstractElement $elementWriter */
+      $elementWriter = new $objectClass($xmlWriter, $element, !$withParagraph);
+      $elementWriter->write();
+      $elements_data .= $xmlWriter->getData();
+    }
+    $blockType = $hasParagraphs ? 'w:p' : 'w:r';
+    $where = $this->findContainingXmlBlockForMacro($search, $blockType);
+    if (is_array($where)) {
+      /** @phpstan-var array{start: int, end: int} $where */
+      $block = $this->getSlice($where['start'], $where['end']);
+      $parts = $hasParagraphs ? $this->splitParagraphIntoParagraphs($block) : $this->splitTextIntoTexts($block);
+      $this->replaceXmlBlock($search, $parts, $blockType);
+      $search = static::ensureMacroCompleted($search);
+      $this->replaceXmlBlock($search, $elements_data, $blockType);
+    }
+  }
+
+  /**
+   * Splits a w:p into a list of w:p where each ${macro} is in a separate w:p.
+   *
+   * @param string $paragraph
+   *
+   * @return string
+   * @throws \CRM_Core_Exception
+   */
+  public function splitParagraphIntoParagraphs(string $paragraph): string {
+    $matches = [];
+    if (1 === preg_match('/(<w:pPr.*<\/w:pPr>)/i', $paragraph, $matches)) {
```

Comment:
Is there a `>` missing before `.*`? The round brackets in the regex seems to be superfluous. Does the regex need the `i` flag?

Reply:
This is adjusted from *PhpWord*'s implementation of [`\PhpOffice\PhpWord\TemplateProcessor::splitTextIntoTexts()`](https://github.com/PHPOffice/PHPWord/blob/8b891bb6842dd383f679b47898fad0b7c181f325/src/PhpWord/TemplateProcessor.php#L1442-L1465). `<w:pPr>` elements seem to be allowed to have attributes, so I think this is correct, although explicitly checking for a space or a `>` would be really correct, if the XSD does allow tag names like `w:pPropoxyphene`.

The `i` flag could be a compatibility thing, but I haven't checked the XSD for whether case variations of tag names are allowed.